homeassistant/helpers/condition.py

```diff
@@ -48,6 +49,10 @@
 
 _LOGGER = logging.getLogger(__name__)
 
+INPUT_ENTITY_ID = re.compile(
+    r"^input_(?:select|text|number|boolean|datetime)\.(?!.+__)(?!_)[\da-z_]+(?<!_)$"
```

Comment:
Do we already have the object id validation as a constant somewhere `'(?!.+__)(?!_)[\da-z_]+(?<!_)'`?

Reply:
I dunno? Not that I'm aware.
This is a descendent from the entity_id validation regex (which I've been using for the HA VSCode extension as well).

Reply:
Maybe something like this in `core.py`

```
DOMAIN_RE = r"(?!.+__)(?!_)[\da-z_]+(?<!_)"
OBJECT_ID_RE = r"(?!_)[\da-z_]+(?<!_)"
VALID_ENTITY_ID = re.compile(r"^" + DOMAIN_RE + r"\." + OBJECT_ID_RE + r"$")
```

and then

`r"^input_(?:select|text|number|boolean|datetime)\." + OBJECT_ID_RE + r"$"`

Reply:
Can an entity be in a `number|boolean|datetime` ?

Reply:
We are not looking for an entity_id, we are looking for a state value of a input_*.

e.g., `states.light.test.state="on"` and `states.input_boolena.light_allowed_on="on"`

This condition will pass:

```yaml
- condition: state
  entity_id: light.test
  state: input_boolean.light_allowed_on
```

Reply:
Gotcha.  We should still be able to reuse the regex pieces though ?

Reply:
```py
DOMAIN_RE = r"(?!.+__)(?!_)[\da-z_]+(?<!_)"
OBJECT_ID_RE = r"(?!_)[\da-z_]+(?<!_)"
VALID_ENTITY_ID = re.compile(r"^{}\.{}$".format(DOMAIN_RE, OBJECT_ID_RE))
```

That is not correct, as the negative lookahead for the double underscore on the domain, needs to be repeated on the object ID (in order to be able to re-use it).

So it becomes this:

```py
DOMAIN_RE = r"(?!.+__)(?!_)[\da-z_]+(?<!_)"
OBJECT_ID_RE = r"(?!.+__)(?!_)[\da-z_]+(?<!_)"
VALID_ENTITY_ID = re.compile(r"^{}\.{}$".format(DOMAIN_RE, OBJECT_ID_RE))
```

The negative lookahead is the most expensive part of the whole regex and this impacts the performance of Home Assistant as a whole.

I think there is a good use-case for the future, but right now, there isn't much to win as it is still localized to a specific situation.

Reply:
Actually I think `fr` works here

Reply:
Suggestion: https://github.com/bdraco/home-assistant/commit/7ca9f866393535be0e6fdc74d1bd268418b19c16

Reply:
Revised suggestion : https://github.com/bdraco/home-assistant/commit/403b5a6ff0e7e9ec5db3642de1227ec159dcce22

Reply:
Admittedly it doesn't really get you much.  It just makes it a bit easier to understand what is going on.

Reply:
Yeah `fr` works really nice in these cases, but this comment still stands: <https://github.com/home-assistant/core/pull/39691#discussion_r484120182>

Reply:
Second suggestion shouldn't suffer from the performance issue: https://github.com/bdraco/home-assistant/commit/403b5a6ff0e7e9ec5db3642de1227ec159dcce22